packages/xod-client/src/hinting/validators.internal.js

```diff
@@ -0,0 +1,486 @@
+      R.defaultTo(defFn),
+      R.prop(R.__, fnMap),
+      R.prop('type')
+    )(action)
+);
+
+export const setOverwritePolicy = errors => ({
+  policy: UPDATE_ERRORS_POLICY.OVERWRITE,
+  errors,
+});
+export const setMergePolicy = errors => ({
+  policy: UPDATE_ERRORS_POLICY.MERGE,
+  errors,
+});
+export const setAssocPolicy = errors => ({
+  policy: UPDATE_ERRORS_POLICY.ASSOC,
+  errors,
+});
+
+// :: [NodeValidateFn] -> [PinValidateFn] -> [LinkValidateFn] -> Project -> Map PatchPath DeducedPinTypes -> Map PatchPath PatchErrors -> [Patch] -> Map PatchPath PatchErrors
+export const validatePatches = R.curry(
+  (
+    nodeValidators,
+    pinValidators,
+    linkValidators,
+    project,
+    allDeducedPinTypes,
+    prevErrors,
+    patches
+  ) =>
+    R.compose(
+      R.map(patch =>
+        mergeAllDeepWithConcat([
+          { errors: {}, nodes: {}, links: {} },
+          getLinkErrors(
+            linkValidators,
+            patch,
+            project,
+            allDeducedPinTypes,
+            prevErrors
+          ),
+          getNodeErrors(
+            nodeValidators,
+            pinValidators,
+            patch,
+            project,
+            prevErrors
+          ),
+        ])
+      ),
+      R.indexBy(XP.getPatchPath)
+    )(patches)
+);
+
+// :: Project -> Map PatchPath DeducedPinTypes -> Map PatchPath PatchErrors -> [Patch] -> Map PatchPath PatchErrors
+export const validatePatchesGenerally = validatePatches(
+  [
+    getDeadRefErrorMap,
+    getTerminalsErrorMap,
+    getVariadicMarkersErrorMap,
+    getAbstractMarkersErrorMap,
+    getConstructorMarkersErrorMap,
+    getBusesErrorMap,
+  ],
+  [validateBoundValues],
+  [validateLinkPins]
+);
+
+// :: Project -> Map PatchPath DeducedPinTypes -> Nullable PatchPath -> Map PatchPath PatchErrors -> Map PatchPath PatchErrors
+export const validateLocalPatches = R.curry(
+  (project, allDeducedPinTypes, changedPatchPath, prevErrors) =>
+    R.compose(
+      validatePatchesGenerally(project, allDeducedPinTypes, prevErrors),
+      filterPatchAndDependentPatchesByPatchPath(changedPatchPath),
+      XP.listLocalPatches
+    )(project)
+);
+
+// :: Project -> Map PatchPath DeducedPinTypes -> Nullable PatchPath -> Map PatchPath PatchErrors -> Map PatchPath PatchErrors
+export const validateAllPatches = R.curry(
+  (project, allDeducedPinTypes, changedPatchPath, prevErrors) =>
+    R.compose(
+      validatePatchesGenerally(project, allDeducedPinTypes, prevErrors),
+      filterPatchAndDependentPatchesByPatchPath(changedPatchPath),
+      XP.listPatches
+    )(project)
+);
+
+// :: Action -> Project -> Map PatchPath DeducedPinTypes -> Map PatchPath PatchErrors -> ErrorsUpdateData
+export const generalValidator = (
+  action,
+  project,
+  allDeducedPinTypes,
+  prevErrors
+) => {
+  const maybePatchPath = getActingPatchPath(action);
+  if (Maybe.isJust(maybePatchPath)) {
+    const patchPath = explodeMaybe('IMPOSSIBLE ERROR', maybePatchPath);
```

Comment:
Quite an ugly construct. Why not `bimap`?

Reply:
`Maybe` is not a bifunctor😔
Possibly, `foldMaybe` will do?